dotCMS/src/main/java/com/dotcms/metrics/RequestTracker.java

```diff
@@ -0,0 +1,278 @@
+     * Get count for specific status code range.
+     * 
+     * @param statusRange the status code range (e.g., "2xx", "4xx")
+     * @return count of responses in that range
+     */
+    public long getStatusCodeCount(String statusRange) {
+        AtomicLong count = statusCodeCounts.get(statusRange);
+        return count != null ? count.get() : 0;
+    }
+    
+    /**
+     * Get count for specific endpoint pattern.
+     * 
+     * @param endpointPattern the endpoint pattern (e.g., "/api/")
+     * @return count of requests to that endpoint pattern
+     */
+    public long getEndpointCount(String endpointPattern) {
+        AtomicLong count = endpointCounts.get(endpointPattern);
+        return count != null ? count.get() : 0;
+    }
+    
+    /**
+     * Calculate overall error rate percentage.
+     * 
+     * @return error rate as percentage (0-100)
+     */
+    public double getErrorRate() {
+        long total2xx = getStatusCodeCount("2xx");
+        long total3xx = getStatusCodeCount("3xx");
+        long total4xx = getStatusCodeCount("4xx");
+        long total5xx = getStatusCodeCount("5xx");
+        
+        long totalRequests = total2xx + total3xx + total4xx + total5xx;
+        long errorRequests = total4xx + total5xx;
+        
+        return totalRequests > 0 ? ((double) errorRequests / totalRequests) * 100 : 0.0;
+    }
+    
+    // ====================================================================
+    // HELPER METHODS
+    // ====================================================================
+    
+    /**
+     * Get status code range category for a status code.
+     * 
+     * @param statusCode the HTTP status code
+     * @return status code range (2xx, 3xx, 4xx, 5xx, other)
+     */
+    private String getStatusCodeRange(int statusCode) {
+        if (statusCode >= 200 && statusCode < 300) return "2xx";
+        if (statusCode >= 300 && statusCode < 400) return "3xx";
+        if (statusCode >= 400 && statusCode < 500) return "4xx";
+        if (statusCode >= 500 && statusCode < 600) return "5xx";
+        return "other";
+    }
+    
+    /**
+     * Get endpoint pattern for a URI.
+     * 
+     * @param uri the request URI
+     * @return endpoint pattern or null if not a tracked pattern
+     */
+    private String getEndpointPattern(String uri) {
+        if (uri == null) return null;
+        
+        if (uri.startsWith("/api/")) return "/api/";
+        if (uri.startsWith("/contentAsset/")) return "/contentAsset/";
+        if (uri.startsWith("/dotAdmin/")) return "/dotAdmin/";
+        if (uri.startsWith("/dotmgt/")) return "/dotmgt/";
+        
+        return null; // Only track key endpoint patterns
+    }
+    
+    /**
+     * Get all status code counts for debugging/monitoring.
+     * 
+     * @return copy of status code counts map
+     */
+    public ConcurrentHashMap<String, Long> getAllStatusCodeCounts() {
+        ConcurrentHashMap<String, Long> result = new ConcurrentHashMap<>();
```

Comment:
A simple hash map looks safe here, isnt?

Reply:
Whether it is safe or not can depend on the use by the caller here and there can be concurrency issues in the use of the returned hashmap.   Saying that though the pattern here is more about immutability rather than concurrency.   We are essentially cloning the values in statusCodeCounts and the correct model would be you use a regular hashmap and then wrap with Collections.unmodifiableMap(result).   This method was really added as a placeholder and example and is not currently being called, so i is probably best to remove for the time being.

Reply:
I have actually fixed up to properly use the methods.